text/0340-firehose-l2.md

```diff
@@ -0,0 +1,924 @@
+writing them to S3.
+
+```ts fixture=with-bucket
+const s3Destination = new destinations.S3(bucket, {
+  prefix: 'myFirehose/DeliveredYear=!{timestamp:yyyy}/anyMonth/rand=!{firehose:random-string}',
+  errorOutputPrefix: 'myFirehoseFailures/!{firehose:error-output-type}/!{timestamp:yyyy}/anyMonth/!{timestamp:dd}',
+});
+```
+
+See: [Custom S3 Prefixes](https://docs.aws.amazon.com/firehose/latest/dev/s3-prefixes.html) in the *Kinesis Data Firehose Developer Guide*.
+
+### Elasticsearch
+
+```ts
+import * as es from '@aws-cdk/aws-elasticsearch';
+import * as destinations from '@aws-cdk/aws-kinesisfirehose-destinations';
+
+const domain = new es.Domain(this, 'Domain', {
+  version: es.ElasticsearchVersion.V7_1,
+});
+
+const deliveryStream = new DeliveryStream(this, 'Delivery Stream', {
+  destinations: [
+    new destinations.Elasticsearch(domain, {
+      indexName: 'myindex',
+    }),
+  ],
+});
+```
+
+### Redshift
+
+A delivery stream can deliver data to a table within a Redshift cluster, using an
+intermediate S3 bucket and executing a Redshift `COPY` command. Redshift clusters must be
+placed in public subnets within an VPC, must be marked as publicly accessible, and cannot
+provide a master user password (it must be generated by the CDK). A Redshift user will be
+created within the cluster for the exclusive use of the delivery stream, and a Redshift
+table with the provided schema will be created within the provided database.
+
+```ts
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as destinations from '@aws-cdk/aws-kinesisfirehose-destinations';
+import * as redshift from '@aws-cdk/aws-redshift';
+import { Duration, Size } from '@aws-cdk/core';
+
+const vpc = new ec2.Vpc(this, 'Vpc');
+const database = 'my_db';
+const cluster = new redshift.Cluster(this, 'Cluster', {
+  vpc: vpc,
+  vpcSubnets: {
+    subnetType: ec2.SubnetType.PUBLIC,
+  },
+  masterUser: {
+    masterUsername: 'master',
+  },
+  defaultDatabaseName: database,
+  publiclyAccessible: true,
+});
+
+const redshiftDestination = new destinations.Redshift(cluster, {
+  user: {
+    username: 'firehose',
+  },
+  database: database,
+  tableName: 'firehose_test_table',
+  tableColumns: [
+    { name: 'TICKER_SYMBOL', dataType: 'varchar(4)' },
+    { name: 'SECTOR', dataType: 'varchar(16)' },
+    { name: 'CHANGE', dataType: 'float' },
+    { name: 'PRICE', dataType: 'float' },
+  ],
+  copyOptions: 'json \'auto\'',
+});
+new DeliveryStream(this, 'Delivery Stream', {
+  destinations: [redshiftDestination],
+});
+```
+
+### 3rd Party
+
+Third-party service providers such as Splunk, Datadog, Dynatrace, LogicMonitor, MongoDB,
+New Relic, and Sumo Logic have integrated with AWS to allow users to configure their
+service as a delivery stream destination out of the box.
+
+These integrations have not been completed (see #1234), please use [custom HTTP endpoints](#custom-http-endpoint)
+to integrate with 3rd party services.
+
+### Custom HTTP Endpoint
```

Comment:
I guess put this before the 3rd party section which says "use this" :)

Reply:
Personally, I don't see the issue with having this located after its been referred to. I'd like to keep the 3rd party section first since each of those destinations should get first-class support in the future and our users should be not be confused about whether they need to use the custom endpoint for these.